### PR TITLE
Week 17 - Project express API - Karolin Albinsson

### PR DESCRIPTION
**Pull request by karolinalbinsson**
frontend/docs : https://karolin-top500albums-frontend.netlify.app/
api call: https://karolin-top-albums.herokuapp.com/

### CODE REVIEW

**Comment by karolinalbinsson** [comment]
Frontend/docs: https://karolin-top500albums-frontend.netlify.app/

File: server.js [hunk 1]
```diff
@@ -1,32 +1,118 @@
...
-// Start defining your routes here
+// Start. return all the endpoints using npm package "express-list-endpoints"
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  if (!res) {
+    res
+      .status(404)
+      .send({ error: 'Error. Try again later.' });
+  } 
+  res.send(myEndpoints(app));
```
Comment (HIPPIEKICK) [inline server.js:22]: Awesome 🙌

File: server.js [hunk 2]
```diff
@@ -1,32 +1,118 @@
...
+  }
+
+  //One album can have multple genres, using includes. 
+  if(filterGenre){
+    albums = albums.filter((item) => item.Genre.toUpperCase().includes(filterGenre.toUpperCase()));
+  }
+
+  //PAGINATION. Limit 50 per request. 
+  //Allow the user to ask for next page by adding parameter. If no parameter is provided, default page is 1. 
+  const albums_paged = albums.slice(page > 1 ? (50*(+page)) : 0, (50*(+page)+(+page === 1 ? 0 :50)));
```
Comment (HIPPIEKICK) [inline server.js:71]: Real nice filtering and pagination!

File: server.js [hunk 3]
```diff
@@ -1,32 +1,118 @@
...
+//get the top 10 albums
+app.get('/api/albums/top10',(req,res) => {
+  const albums = albumData.slice(0,10);
+  if(albums.length < 1){
+    res.status(404).send({
+      error: 'Not found, please try later.'
+    })
+  }
+  res.json(albums);
+})
```
Comment (HIPPIEKICK) [inline server.js:98]: This could also be a query under your main endpoint

File: server.js [hunk 4]
```diff
@@ -1,32 +1,118 @@
...
+
+//SINGLE ITEM
+//Get album based on placement on list using params
+app.get('/api/albums/placement/:placement',(req,res) => {
+  const placement = req.params.placement;
+  const album = albumData.find((item) => item.Number === +placement);
+  album ? res.json(album) : res.status(404).send({ error: `No album with placement ${placement} found.` });
+})
```
Comment (HIPPIEKICK) [inline server.js:106]: Clever!

File: server.js [hunk 5]
```diff
@@ -1,32 +1,118 @@
...
+
+//Get album based on title using params
+app.get('/api/albums/title/:title',(req,res) => {
+  const title = req.params.title.replaceAll('+',' ');
+  const album = albumData.find((item) => item.Album.toString().toUpperCase().includes(title.toUpperCase()));
+  album ? res.json(album) : res.status(404).send({ error: `No album with title ${title} found.` });
+})
```
Comment (HIPPIEKICK) [inline server.js:113]: This too could be a query under your main endpoint, something like:
`/albums?title=pet%20sounds`

**Review by HIPPIEKICK — APPROVED** [review]
Real nice frontend and impressive filtering. Keep it up 💯

**Pull request closed without merging**